model/generate.go

```diff
@@ -1,25 +1,296 @@
+type GeneratedProject struct {
+	BuildVariants []parserBV                 `yaml:"buildvariants"`
+	Tasks         []parserTask               `yaml:"tasks"`
 	Functions     map[string]*YAMLCommandSet `yaml:"functions"`
+
+	TaskID string
+}
+
+// MergeGeneratedProjects takes a slice of generated projects and returns a single, deduplicated project.
+func MergeGeneratedProjects(projects GeneratedProjects) *GeneratedProject {
+	catcher := grip.NewBasicCatcher()
+
+	bvs := map[string]*parserBV{}
+	tasks := map[string]*parserTask{}
+	functions := map[string]*YAMLCommandSet{}
+
+	for _, p := range projects {
+		for _, bv := range p.BuildVariants {
+			if len(bv.Tasks) == 0 {
+				if _, ok := bvs[bv.Name]; ok {
+					catcher.Add(errors.Errorf("found duplicate buildvariant (%s)", bv.Name))
+				} else {
+					bvs[bv.Name] = &bv
+				}
+			}
+			if _, ok := bvs[bv.Name]; ok {
+				bvs[bv.Name].Tasks = append(bvs[bv.Name].Tasks, bv.Tasks...)
+			}
+			bvs[bv.Name] = &bv
+		}
+		for _, t := range p.Tasks {
+			if _, ok := tasks[t.Name]; ok {
+				catcher.Add(errors.Errorf("found duplicate task (%s)", t.Name))
+			} else {
+				tasks[t.Name] = &t
+			}
+		}
+		for f, val := range p.Functions {
+			if _, ok := functions[f]; ok {
+				catcher.Add(errors.Errorf("found duplicate function (%s)", f))
+			}
+			functions[f] = val
+		}
+	}
+
+	g := &GeneratedProject{}
+	for i := range bvs {
+		g.BuildVariants = append(g.BuildVariants, *bvs[i])
+	}
+	for i := range tasks {
+		g.Tasks = append(g.Tasks, *tasks[i])
+	}
+	g.Functions = functions
+	return g
 }
 
-// GeneratedTasksFromJSON returns a GeneratedTasks type from JSON. We use the
+// ParseProjectFromJSON returns a GeneratedTasks type from JSON. We use the
 // YAML parser instead of the JSON parser because the JSON parser will not
 // properly unmarshal into a struct with multiple fields as options, like the YAMLCommandSet.
-func GeneratedTasksFromJSON(data []byte) (*GeneratedTasks, error) {
-	g := &GeneratedTasks{}
-	if err := yaml.Unmarshal(data, g); err != nil {
-		return nil, errors.Wrap(err, "error unmarshaling into GeneratedTasks")
+func ParseProjectFromJSON(data []byte) (GeneratedProject, error) {
+	g := GeneratedProject{}
+	if err := yaml.Unmarshal(data, &g); err != nil {
+		return g, errors.Wrap(err, "error unmarshaling into GeneratedTasks")
 	}
 	return g, nil
 }
+
+// AddGeneratedProjectToVersion adds the buildvariants, tasks, and functions
+// from a generated project config to a project.
+func (g *GeneratedProject) AddGeneratedProjectToVersion() error {
+	// Get task, version, and project.
+	t, err := task.FindOneId(g.TaskID)
+	if err != nil {
+		return errors.Wrapf(err, "error finding task %s", g.TaskID)
+	}
+	v, err := version.FindOneId(t.Version)
+	if err != nil {
+		return errors.Wrapf(err, "error finding version %s", t.Version)
+	}
+	p := &Project{}
+	if err := LoadProjectInto([]byte(v.Config), t.Project, p); err != nil {
+		return errors.Wrap(err, "error reading project yaml")
+	}
```

Comment:
I think I recently added a method for "find project config for a task" method, but if not, perhaps we should refactor this into a more generic method.

Reply:
we use the task and version later though, so in this case a refactor would not be useful